pkl-core/src/main/java/org/pkl/core/runtime/VmUtils.java

```diff
@@ -281,44 +274,53 @@ public static Object doReadMember(
             }
           } catch (VmException e) {
             CompilerDirectives.transferToInterpreter();
-            // A failed property type check looks as follows in the stack trace:
-            // x: Int(isPositive)
-            // at ...
-            // x = -10
-            // at ...
-            // However, if the value of `x` is a parse-time constant (as in `x = -10`),
-            // there's no root node for it and hence no stack trace element.
-            // In this case, insert a stack trace element to make the stack trace look the same
-            // as in the non-constant case. (Parse-time constants are an internal optimization.)
-            e.getInsertedStackFrames()
-                .put(
-                    callTarget,
-                    createStackFrame(member.getBodySection(), member.getQualifiedName()));
+            insertStackFrame(member, callTarget, e);
             throw e;
           }
         }
-      } else if (receiver instanceof VmListingOrMapping<?> vmListingOrMapping) {
-        if (owner != receiver && owner instanceof VmListingOrMapping<?> vmListingOrMappingOwner) {
-          ret = vmListingOrMappingOwner.typecastObjectMember(member, ret, callNode);
-        }
-        ret = vmListingOrMapping.typecastObjectMember(member, ret, callNode);
+        // `if (receiver instanceof VmListingOrMapping)` doesn't work
+        // (only) because PropertiesRenderer amends a VmDynamic with a VmListing (hack?)
+      } else if (owner instanceof VmListingOrMapping) {
+        result =
+            ((VmListingOrMapping) receiver)
+                .doTypeCast(constantValue, owner, callNode, member, null);
```

Comment:
Agreed with "better to remove the hack" above, but this reads weird, even with the comment, maybe...
```suggestion
      } else if (owner instanceof VmListingOrMapping && receiver instance VmListingOrMapping checkable) {
        result = checkable.doTypeCast(constantValue, owner, callNode, member, null);
```

Reply:
changed to (almost) your code